staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go

```diff
@@ -101,62 +119,200 @@ func WithAudit(handler http.Handler, requestContextMapper request.RequestContext
+			if longRunning {
+				// synchronize with the go routine. If a status code has been written, the
+				// go routine above will get the first chance to send an audit event for it.
+				close(codeCh)
+				<-firstEventSentCh
+			}
+
+			if r := recover(); r != nil {
+				ev.ResponseStatus = &metav1.Status{
+					Code: http.StatusInternalServerError,
+				}
+				if err := backend.LogAuditEvent(ev); err != nil {
+					glog.Errorf(err.Error())
+				}
+				panic(r)
+			}
+
+			if ev.ResponseStatus == nil {
+				ev.ResponseStatus = &metav1.Status{
+					Code: 200,
+				}
+			}
+
+			select {
+			case code, ok := <-codeCh:
+				if ok {
+					ev.ResponseStatus.Code = int32(code)
+				}
+			default:
+			}
+
+			if err := backend.LogAuditEvent(ev); err != nil {
+				glog.Errorf(err.Error())
+			}
+		}()
 		handler.ServeHTTP(respWriter, req)
 	})
 }
 
-func auditStringSlice(inList []string) string {
-	quotedElements := make([]string, len(inList))
-	for i, in := range inList {
-		quotedElements[i] = fmt.Sprintf("%q", in)
+func decorateResponseWriter(responseWriter http.ResponseWriter, firstEventSentCh <-chan struct{}) (http.ResponseWriter, chan int) {
+	codeCh := make(chan int, 1)
+	delegate := &auditResponseWriter{
+		ResponseWriter:   responseWriter,
+		code:             codeCh,
+		firstEventSentCh: firstEventSentCh,
 	}
-	return strings.Join(quotedElements, ",")
-}
 
-func decorateResponseWriter(responseWriter http.ResponseWriter, out io.Writer, id string) http.ResponseWriter {
-	delegate := &auditResponseWriter{ResponseWriter: responseWriter, out: out, id: id}
 	// check if the ResponseWriter we're wrapping is the fancy one we need
 	// or if the basic is sufficient
 	_, cn := responseWriter.(http.CloseNotifier)
 	_, fl := responseWriter.(http.Flusher)
 	_, hj := responseWriter.(http.Hijacker)
 	if cn && fl && hj {
-		return &fancyResponseWriterDelegator{delegate}
+		return &fancyResponseWriterDelegator{delegate}, codeCh
 	}
-	return delegate
+	return delegate, codeCh
+}
+
+var _ http.ResponseWriter = &auditResponseWriter{}
+
+// auditResponseWriter intercepts WriteHeader and sends the code through the code channel. Because this blocks,
+// the audit code has time to send the event without risking concurrent writes. WriteHeader will continue when
+// firstEventSentCh has been closed after the audit event has been sent out.
+type auditResponseWriter struct {
+	http.ResponseWriter
+
+	// capacity 1, i.e. non-blocking
+	sendCodeOnce     sync.Once
+	code             chan<- int
+	firstEventSentCh <-chan struct{}
+}
+
+func (a *auditResponseWriter) sendCode(code int) {
+	a.sendCodeOnce.Do(func() {
+		// this does not block
+		a.code <- code
+
+		if a.firstEventSentCh != nil {
+			// this blocks until a.firstEventSentCh is closed
+			<-a.firstEventSentCh
+		}
+	})
+}
+
+func (a *auditResponseWriter) Write(bs []byte) (int, error) {
+	a.sendCode(200) // the Go library call WriteHeader internally if no code was written yet. But this will go unnoticed for us
```

Comment:
net/http.response sets 200 in `Write` explicitly if no `WriteHeader` call was seen. Moreover, it ignored later `WriteHeader` calls.

In order to create the first event on Write-without-WriteHeader, we need this call.